File: Arrays/kadanes-algorithm/maximum-subarray-sum.js

// Brute Force Approach
const maximumSubArraySum = (nums) => {
    const n = nums.length;

    let maxSum = -Infinity;
    
    for(let start = 0; start < n; start++) { // starting point
        for(let end = start; end < n; end++){ // ending point
            let currSum = 0;
            for(let i = start; i <= end; i++) {
                currSum += nums[i];
                maxSum = Math.max(currSum, maxSum);
            }
        }
    }
    console.log(maxSum); // 21
}


// Optimal Approach - Kadane's Algorithm
const maximumSubArraySumOptimal = (nums) => {
  let maxSum = -Infinity;
  let currSum = 0;
  const n = nums.length;
  for (let i = 0; i < n; i++) {
    currSum += nums[i];
    maxSum = Math.max(currSum, maxSum);
    if (currSum < 0) {
      currSum = 0;
    }
  }
  console.log(maxSum); //// 21
}

function main() {
    const nums = [-1, 8, 4, 5, 0, -3, 7]
    maximumSubArraySum(nums); 
    maximumSubArraySumOptimal(nums); 
}

main();

File: Arrays/kadanes-algorithm/MaximumSubArraySum.java

public class MaximumSubArraySum {

    public static void main(String[] args) {
        int[] nums = {3, -4, 5, 20, -10, 7, 8};
        maximumSubArraySumOptimal(nums);
        maximumSubArraySum(nums);
    }


    // Brute - Force Approach
    public static void maximumSubArraySum(int[] nums) {
        int maxSum = Integer.MIN_VALUE;
        int n = nums.length;
        for(int start = 0; start < n; start++) { // starting point
            for(int end = start; end < n; end++) { // ending point
                int currSum = 0;
                for(int i = start; i <= end; i++) {
                    currSum += nums[i];
                }
                maxSum = Math.max(maxSum, currSum);
            }
        }
        System.out.println(maxSum); // 30
    }


    // optimal approach - kadane's algorithm
    public static void maximumSubArraySumOptimal(int[] nums) {
        int n = nums.length;
        int currSum = 0;
        int maxSum = Integer.MIN_VALUE;

        for(int i = 0; i < n; i++) {
            currSum += nums[i];
            maxSum = Math.max(maxSum, currSum);
            if(currSum < 0) {
                currSum = 0;
            }
        }
        System.out.println(maxSum); // 30
    }
}
